src/udev/udev-builtin-input_id.c

```diff
@@ -262,6 +267,30 @@ static bool test_pointers(sd_device *dev,
         if (is_mouse && id->bustype == BUS_I2C)
                 is_pointing_stick = true;
 
+        /* Joystick un-detection. Some keyboards have random joystick buttons
+         * set. Avoid those being labeled as ID_INPUT_JOYSTICK with some heuristics.
+         * The well-known keys represent a (randomly picked) set of key groups.
+         * A joystick may have one of those but probably not several. And a joystick with less than 2 buttons
+         * or axes is not a joystick either.
+         * libinput uses similar heuristics, any changes here should be added to libinput too.
+         */
+        if (is_joystick) {
+                static const unsigned int well_known_keyboard_keys[] = {
+                        KEY_LEFTCTRL, KEY_CAPSLOCK, KEY_NUMLOCK, KEY_INSERT,
+                        KEY_MUTE, KEY_CALC, KEY_FILE, KEY_MAIL, KEY_PLAYPAUSE,
+                        KEY_BRIGHTNESSDOWN,
+                };
+                size_t num_well_known_keys = 0;
+
+                for (size_t i = 0; has_keys && i < ELEMENTSOF(well_known_keyboard_keys); i++) {
+                        if (test_bit(well_known_keyboard_keys[i], bitmask_key))
+                                num_well_known_keys++;
+                }
```

Comment:
`has_keys` is not changed in the loop, hence it is not necessary to evaluated every time. Also, `{}` can be dropped:
```
        if (has_keys)
                for (size_t i = 0; i < ELEMENTSOF(well_known_keyboard_keys); i++)
                        if (test_bit(well_known_keyboard_keys[i], bitmask_key))
                                num_well_known_keys++;
```

Reply:
fixed, thanks